core/views.py

```diff
@@ -148,50 +183,77 @@ def setting(request):
     )
 
 
-@login_required
 def save_hackathon(request: HttpRequest, hackathon_id):
     if request.method == "POST":
         hackathon = get_object_or_404(Hackathon, id=hackathon_id)
         hacker = get_object_or_404(Hacker, id=request.user.id)
+        page_type = request.GET.get("page_type")
 
         if hacker.saved.filter(id=hackathon_id).exists():
+            hacker.saved.remove(hackathon)
+            if page_type == "saved":
+                return redirect("saved_hackathons")
+            else:
+                return JsonResponse({"status": "sucess", "hackathon": "removed"})
+
+        hacker.saved.add(hackathon)
+        if page_type == "saved":
+            return redirect("saved_hackathons")
+        else:
             return JsonResponse(
-                {"status": "fail", "error": "Invalid request"}, status=400
+                {
+                    "status": "success",
+                    "hackathon": {
+                        "id": hackathon.id,
+                        "name": hackathon.name,
+                    },
+                }
             )
 
-        hacker.saved.add(hackathon)
-        return JsonResponse(
-            {
-                "status": "success",
-                "hackathon": {
-                    "id": hackathon.id,
-                    "name": hackathon.name,
-                },
-            }
-        )
 
+def str_to_bool(s):
+    return s.lower() in ["true", "1", "yes"]
 
-@login_required
-def unsave_hackathon(request: HttpRequest, hackathon_id):
+
+def add_vote(request: HttpRequest, hackathon_id):
     if request.method == "POST":
         hackathon = get_object_or_404(Hackathon, id=hackathon_id)
         hacker = get_object_or_404(Hacker, id=request.user.id)
+        type_vote = request.GET.get("vote_state")
+        state = str_to_bool(type_vote)
+        vote = Vote.objects.filter(hackathon_id=hackathon, from_hacker=hacker)
+
+        if vote.exists():
```

Comment:
checking if .exists then .first causes an n+1 query. just run .first and check if None